### ISSUE TITLE
thd_windows.cc failed to build on Windows clang when build target is 32bit

### ISSUE THREAD

**Issue by equinox1993** [grpc/grpc#18083]
### What version of gRPC and what language are you using?
 Commit at 872ca1f8a51ff41e2f3972d0d64b0e34461d639e
 
### What operating system (Linux, Windows, …) and version?
 Windows 10 Version 1709 (OS Build 16299.904)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 clang version 9.0.0 (trunk 353250)
 Target: x86_64-pc-windows-msvc
 Thread model: posix
 
### What did you do?
 Build grpc++ with the architecture set to 32bit.
 
### What did you expect to see?
 Library successfully builds.
 
### What did you see instead?
 Failed to build the `thd_windows.cc` file, because clang cannot convert a normal lambda into a stdcall function.

```
../../third_party/grpc/src/src/core/lib/gprpp/thd_windows.cc /Foobj/third_party/grpc/grpc_core/thd_windows.obj /Fd"obj/third_party/grpc/grpc_core_cc.pdb"
../../third_party/grpc/src/src/core/lib/gprpp/thd_windows.cc(74,16):  error: no matching function for call to 'CreateThread'
      handle = CreateThread(
               ^~~~~~~~~~~~
..\..\third_party\depot_tools\win_toolchain\vs_files\e04af53255fe13c130e9cfde7d9ac861b9fb674a\win_sdk\Include\10.0.17763.0\um\processthreadsapi.h(170,1):  note: candidate function not viable: no known conversion from '(lambda at ../../third_party/grpc/src/src/core/lib/gprpp/thd_windows.cc:76:11)' to 'LPTHREAD_START_ROUTINE' (aka 'unsigned long (*)(void *) __attribute__((stdcall))') for 3rd argument
CreateThread(
^
```

https://github.com/grpc/grpc/blob/872ca1f8a51ff41e2f3972d0d64b0e34461d639e/src/core/lib/gprpp/thd_windows.cc#L76

Looks like clang pretends to be MSVC and sets the `_MSC_VER` macro, which causes `WIN_LAMBDA` to be defined as empty and the lambda here to be a normal lambda. Clang can't automatically convert lambda to match the calling convention AFAICT, so it failed to build.

I don't think clang supports tagging a lambda for calling convention. One possible fix is to move the code inside the lambda into a private static method and tag the method with `WINAPI`.
 
### Anything else we should know about your project / environment?

I'm trying to add grpc into Chromium ([CL](https://chromium-review.googlesource.com/c/chromium/src/+/1444971)) and this bug makes some Chromium trybots fail to build.

**Comment by vjpai** [comment]
Cc @nicolasnoble @jtattermusch : I discussed this offline with @equinox1993 and the reporter is planning to send in a PR to fix this problem . This issue is interesting since this is a user case wanting to use clang with Windows. Is this something that we need to support in our CI?

**Comment by equinox1993** [comment]
FYI I seem to be able to run `run_tests.py` with clang-cl. Here are what I did:

1. [Install LLVM](http://releases.llvm.org/download.html)
1. Install the [LLVM compiler toolchain](https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.llvm-toolchain) on visual studio
1. Modify `pre_build_cmake.bat` to add `-T LLVM` to the cmake flag.
1. Run `python tools/run_tests/run_tests.py -l c++ --build_only`

Then I can reproduce the the error I saw when building on Chromium:

```
"C:\src\grpc\cmake\build\buildtests_cxx.vcxproj" (default target) (1) ->
"C:\src\grpc\cmake\build\address_sorting_test.vcxproj" (default target) (3) ->
"C:\src\grpc\cmake\build\gpr.vcxproj" (default target) (45) ->
(ClCompile target) ->
  C:\src\grpc\src\core\lib\gprpp\thd_windows.cc(74): error : no matching function for call to 'CreateThread' [C:\src\grpc\cmake\build\gpr.vcxproj]
```

I'm also seeing some other warnings/errors that I didn't see on Chromium. Maybe Chromium has suppressed some warnings for third party code and the GN build file I wrote didn't build some of the files that failed to build.

**Comment by nicolasnoble** [comment]
I'm not really eager to add yet another blocking edge case scenario in our PRs, no. I would personally deem that only Visual Studio is our main supported compiler, and that other cases is supported on best effort, as we do with many of our edge case scenarios.

It is true that Visual Studio now has added clang/llvm support so that people can now easily build their code using clang/llvm using Visual Studio directly, but the installation procedure is still painful, and I'm still not seeing indicators this usage is more prevalent than Microsoft's compiler in general.

Bottom line is: we can't do first-class compiler/platform support for every single combinations out there. 

That being said, since this is about another Google product, I'm open to discussion.

**Comment by equinox1993** [comment]
We don't plan to roll dependency on gRPC very often unless there is blocking bug or security concern, so I think it's fine not to run CI for clang+Windows. I'm happy to fix bug on gRPC if it blocks development on Chromium.

Also I think it would be great if `run_tests.py` has an option to build on clang+Windows. This can be done by simply adding a `-T LLVM` flag to the cmake parameter in addition to the existing `-G "Visual Studio 14 2015"` flag. This can make testing easier for us.

**Comment by nicolasnoble** [comment]
Yeah, that last part is fine and good to have. I'd like to point out that we'd rather move towards bazel as a test framework however. If this can be achieved using bazel instead, I'd be even more supportive.

**Comment by equinox1993** [comment]
Looks like this PR makes it possible to use clang-cl on Bazel by setting a `USE_CLANG_CL` environment variable: https://github.com/bazelbuild/bazel/pull/6553

So we probably don't need to do anything on gRPC to support that :)

**Comment by jtattermusch** [comment]
AFAIK our only supported compiler on Windows is MSVC. Other compilers are supported on best effort basis (and we don't have the capacity to test them continuously and monitor the test results).
We're fine with accepting a one-time fix for your problem, but I don't want to add more stuff to run_tests.py.
